src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
+
+                Buffer.BlockCopy(array, offset, _buffer, _writePos, remainingBuffer);
+
+                byte[] oldBuffer = _buffer;
+                _buffer = new byte[_bufferSize];
```

Comment:
That's true. Let me think about this a bit over the weekend. This would all get much simpler if we said that all async operations are serialized and you can't do synchronous operations while an async operation is in flight. Then we shouldn't need any additional buffer allocations.